x/trigger/keeper/trigger_dispatcher.go

```diff
@@ -0,0 +1,97 @@
+package keeper
+
+import (
+	"fmt"
+
+	errorsmod "cosmossdk.io/errors"
+
+	types "github.com/cosmos/cosmos-sdk/codec/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdktx "github.com/cosmos/cosmos-sdk/types/tx"
+	"github.com/cosmos/cosmos-sdk/x/group/errors"
+)
+
+const (
+	MaximumActions  uint64 = 5
+	MaximumQueueGas uint64 = 2000000
+)
+
+// ProcessTriggers Reads triggers from queues and attempts to run them.
+func (k Keeper) ProcessTriggers(ctx sdk.Context) {
+	var actionsProcessed uint64
+	var gasConsumed uint64
+
+	for !k.QueueIsEmpty(ctx) && actionsProcessed < MaximumActions {
+		item := k.QueuePeek(ctx)
+		triggerID := item.GetTrigger().Id
+		gasLimit := k.GetGasLimit(ctx, triggerID)
+
+		if gasLimit+gasConsumed > MaximumQueueGas {
+			return
+		}
+		actionsProcessed++
+		gasConsumed += gasLimit
+
+		k.Dequeue(ctx)
+		k.RemoveGasLimit(ctx, triggerID)
+
+		actions := item.GetTrigger().Actions
+		k.runActions(ctx, gasLimit, actions)
+	}
+}
+
+// RunActions Runs all the actions and constrains them by gasLimit.
+func (k Keeper) runActions(ctx sdk.Context, gasLimit uint64, actions []*types.Any) {
+	cacheCtx, flush := ctx.CacheContext()
+	gasMeter := sdk.NewInfiniteGasMeter()
+	cacheCtx = cacheCtx.WithGasMeter(gasMeter)
+
+	msgs, err := sdktx.GetMsgs(actions, "RunActions")
+	if err != nil {
+		k.Logger(ctx).Error(
+			"GetMsgs",
+			"actions", actions,
+			"error", err,
+		)
+		return
+	}
+	results, err := k.handleMsgs(cacheCtx, msgs, gasLimit)
+	if err != nil {
+		k.Logger(ctx).Error(
+			"HandleMsgs",
+			"error", err,
+		)
+		return
+	}
```

Comment:
I feel like this error should be recorded somewhere more accessible than node logs.

One option is to store it in state. E.g. have a "trigger results" key prefix. That gets a little tricky here since the result needs to be persisted even if not calling `flush()`. We'd also want a "trigger results" query.

Another option is to emit an event with the trigger id and error. That way it's at least not lost.

Reply:
Are you sure we would want an error in an event? I could definitely see having an event like TriggerExecutionEvent, and a flag for if it was successful. This does bring up an interesting point though about seeing the results. The only way currently is by seeing a successful transaction emit events.

Reply:
Since these are being executed outside of any Tx, if there _is_ an error, there's no way for the user to find out. I guess, maybe we should check what the gov module does for those. I know it's a pain to research failed gov props, but I'm pretty sure it's possible without having node logs. And having it be possible is better than not having it.

The only two places I can think to propagate such errors is either in state or in the events.

And I just looked at what the gov module does. They just mark the proposal status as "failed" and write to the logs. The actual error isn't stored anywhere. :/

So maybe we punt this down the road as a possible enhancement. I agree that it makes me nervous to try to put an error message in state.